Add Notifications CMS functionality

jschultze opened the pull request:
author Johannes Schultze <schultze@effective-webwork.de> 1695107528 +0200 committer Johannes Schultze <schultze@effective-webwork.de> 1743752864 +0200

Added "notifications"-CMS functionality.

Added composer packages for emoji support in notifications.

Update of generated files.

Broadcasts display the contents of the default language if the currently selected language of the user is not available.

jschultze committed:
parent e6f124953b80536cca3bbace988564a4841f64b2
author Johannes Schultze <schultze@effective-webwork.de> 1695107528 +0200
committer Johannes Schultze <schultze@effective-webwork.de> 1743752864 +0200

Added "notifications"-CMS functionality.

Added composer packages for emoji support in notifications.

Update of generated files.

Broadcasts display the contents of the default language if the currently selected language of the user is not available.

jschultze commented:
Continuing work from https://github.com/vufind-org/vufind/pull/3103

jschultze committed:
Fixed the display of broadcasts and pages in the default language if not local translation is available.

jschultze commented:
> We think this should be set up in a different way, so that the first logical option is to display notifications in the language that matches the language the user sets for the page, and then if the user picks a different language, the banner should still be displayed in the language it was written in.

The notification system was supposed to behave just like this. The error is fixed now.

jschultze commented:
@sturkel89 @demiankatz Regarding the wrapping: I can reproduce the problem when using quite long texts as the "Navigation title". I tried several approaches with CSS to shorten the title, and tried moving the `nav-item`-li tags, but with no success. The same wrapping would occur, if the texts for the login-button, etc. where much longer.

I will give this to a colleague who is more experienced in working with CSS, but I'm not sure if there's a really good fix for this (next to keeping the text short).

jschultze commented:
@sturkel89 

> The broadcast banners have an X in the upper right-hand corner. When I mouse over the X, the cursor becomes a pointing finger, indicating that it might be able to affect things. But when I try to click any of the Xes, nothing happens. You say that you can click and the banner disappears for the session. I would like to see that behavior, but I'm not seeing it on Chrome or Firefox on Windows.

Can you take a look at the developer console in the browser and see if any errors occur?

demiankatz commented on composer.json:
```diff
@@ -100,7 +101,9 @@
         "wikimedia/composer-merge-plugin": "2.1.0",
-        "yajra/laravel-pdo-via-oci8": "3.4.4"
+        "yajra/laravel-pdo-via-oci8": "3.4.4",
+        "league/commonmark": "2.3.9",
+        "elgigi/commonmark-emoji": "2.0.0"
```
I think these lines should be removed -- they seem to be duplicates of dependencies already listed above (and in the case of league/commonmark, an older version).

jschultze replied:
The lines are removed.

demiankatz reviewed (changes requested):
See below for a small problem that I noticed when taking a quick glance at the diffs here.

demiankatz commented:
> @sturkel89 @demiankatz Regarding the wrapping: I can reproduce the problem when using quite long texts as the "Navigation title". I tried several approaches with CSS to shorten the title, and tried moving the `nav-item`-li tags, but with no success. The same wrapping would occur, if the texts for the login-button, etc. where much longer.
> 
> I will give this to a colleague who is more experienced in working with CSS, but I'm not sure if there's a really good fix for this (next to keeping the text short).

If your colleague doesn't come up with any suggestions, we could also check with @crhallberg for ideas. Or, as you say, maybe enforcing some kind of length limit (or at least warning past a certain threshold) would encourage behavior that avoids the problem.

jschultze committed:
Fixed the display of pages links in the header menu.

jschultze commented:
@sturkel89 I think we found a solution for the wrapping. The display of the ellipsis start quite early, but the might encourage admins to shorten the navigation title.

jschultze committed:
Removed duplicate entries in composer json.

jschultze committed:
Added database service for notifications pages.

jschultze commented:
@demiankatz I added the database service structure for pages and will continue to add the same for broadcasts, probably not today though.

sturkel89 commented:
Hi Johannes - I can't get into the PR to view it or try things with the developer console open because I'm getting a "Division by zero" error when in development mode.

Demian noticed this in the code: 

  `$pages = $this->notifications()->getPages();`
  `$maxWidth = round((100/sizeof($pages))-10);`
  
Perhaps you can view the PR because you already have notifications set up, but since I am starting fresh with no notifications I cannot get in.

demiankatz commented on module/VuFind/src/VuFind/Db/Row/Pages.php:
```diff
@@ -0,0 +1,64 @@
+<?php
+
+/**
+ * Row Definition for broadcasts
```
```suggestion
 * Row Definition for pages
```

jschultze replied:
Will be fixed in next commit.

demiankatz commented on module/VuFind/src/VuFind/Db/Row/Pages.php:
```diff
@@ -0,0 +1,64 @@
+use VuFind\Db\Entity\PagesEntityInterface;
+
+/**
+ * Row Definition for index files
```
```suggestion
 * Row Definition for pages
```

jschultze replied:
Will be fixed in next commit.

demiankatz commented on module/VuFind/src/VuFind/Db/Entity/PagesEntityInterface.php:
```diff
@@ -0,0 +1,54 @@
+     * @return int|null
+     */
+
+    public function getPrimaryKeyId(): ?int;
```
This interface should have getters and setters for all of the fields in the notifications_pages table. It would also make sense for it to be named NotificationsPagesInterface for consistency with the table name, and to differentiate from any other kinds of pages that might show up in the database in the future.

I'm not sure what `getPrimaryKeyId` is for, but if this is an internal Laminas thing that I'm forgetting about, then it definitely should _not_ be part of the interface.

(The purpose of the interface is to define all data fields, and create setters and getters to access them; when we migrate to Doctrine, using magic properties representing the field names will no longer be an option).

jschultze replied:
I think it is a technical id that is returned after a new entry in the database is created. This id has to be set for all broadcasts or pages objects as the page_id, to connect versions of different languages.

I will add setters and getters.

demiankatz replied:
Maybe I'm misunderstanding something, but I think your `getPrimaryKeyId` method will return the same value as `getId`, it's just pulling it from a different place. I think it would be good to eliminate this if it's truly redundant, but I'm happy to discuss further or look more carefully if I'm mistaken.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/PagesService.php:
```diff
@@ -0,0 +1,163 @@
+    /**
+     * Constructor
+     *
+     * @param Pages $pages Session table object
```
```suggestion
     * @param Pages $pages Notification pages table object
```

jschultze replied:
Will be fixed in next commit.

demiankatz commented on module/VuFind/src/VuFind/Db/Table/Pages.php:
```diff
@@ -0,0 +1,304 @@
+     */
+    public function getPageById($id)
+    {
+        if ($id) {
```
Do you need this if clause?

jschultze replied:
Will be fixed in next commit.

demiankatz commented on module/VuFind/src/VuFind/Db/Table/Pages.php:
```diff
@@ -0,0 +1,304 @@
+     */
+    public function getPagesByPageId($page_id)
+    {
+        if ($page_id) {
```
Do you need this if clause?

jschultze replied:
Will be fixed in next commit.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/PagesServiceInterface.php:
```diff
@@ -0,0 +1,125 @@
+     *
+     * @return mixed Array of page objects
+     */
+    public function getPagesByPageId(int $page_id): mixed;
```
Should this type be `array` rather than `mixed`? Because of the if clause I asked about, it seems it can potentially return null, but should we return empty array when no valid data is found so that the types are more predictable?

jschultze replied:
Will be changed to array with the next commit.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/PagesServiceInterface.php:
```diff
@@ -0,0 +1,125 @@
+    /**
+     * Get a page object by id
+     *
+     * @param int $id Id of the page
```
It seems like there are two different kinds of IDs -- a plain primary key ID used by this method, and a "page ID" used by the following method. It's a little confusing that both of these methods have identical parameter descriptions. Should these comments be clarified somehow?

jschultze replied:
We need both ids to connect versions of different languages. The comment will be fixed with the next commit.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/PagesService.php:
```diff
@@ -0,0 +1,163 @@
+     * @param int $index   New position of the page
+     * @param int $page_id Id of the page
+     */
+    public function setPriorityForPageId($index, $page_id): void
```
Would it make sense to use a `static` return type, and return `$this` from the method, to implement a fluent interface here? Or is that not helpful?

Either way, the interface return types should be adjusted to match the class return types (currently the setters' void returns are missing) and the comments are missing `@return` annotations.

jschultze replied:
The return types will be adjusted with the next commit.

demiankatz commented on module/VuFind/src/VuFind/Db/Service/PagesService.php:
```diff
@@ -0,0 +1,163 @@
+     * @param int $visibility_global New visibility of the page
+     * @param int $page_id           Id of the page
+     */
+    public function setVisibilityGlobalForPageId($visibility_global, $page_id): void
```
Should this be `setGlobalVisibilityForPageId` rather than `setVisibilityGlobalForPageId`?

jschultze replied:
I ws following the name of the variables (`visibility` and `visibiliy_global`) throughout the code.

demiankatz replied:
Ahh, I see, that makes sense!

demiankatz reviewed (changes requested):
Thanks for the progress, @jschultze -- I took a relatively quick glance at it just to see how things are going, and I left a few comments below.

jschultze commented:
@demiankatz

> It would also make sense for it to be named NotificationsPagesInterface for consistency with the table name, and to differentiate from any other kinds of pages that might show up in the database in the future.

Should all Classes in `/Vufind/Db/...` be renamed to `Notifications...`?

demiankatz commented:
> Should all Classes in `/Vufind/Db/...` be renamed to `Notifications...`?

I think that would make sense for consistency.

jschultze committed:
Refactoring of Notifications class names.

jschultze committed:
Fixed a possible division by zero.

jschultze committed:
Added base classes for NotificationsBroadcasts entity and service.

jschultze committed:
Fixed comments.

jschultze committed:
Fixed comments.

jschultze committed:
Added setters and getters.

jschultze commented:
@demiankatz I fixed the problems listed here and renamed all classes throughout the Code with "Notifications...". I added the setters and getters to the Notifications...EntityInterface-classes, but I'm not sure right now, how much the code in the row and table classes should be refactored to the new model, as the code does not work at the moment (Class VuFind\Db\Row\NotificationsPages contains 26 abstract methods and must therefore be declared abstract or implement the remaining methods (VuFind\Db\Entity\NotificationsPagesEntityInterface::setId, VuFind\Db\Entity\NotificationsPagesEntityInterface::getId, VuFind\Db\Entity\NotificationsPagesEntityInterface::setPageId, ...)).

demiankatz commented:
@jschultze, yes, it is necessary to implement all of the getters and setters on the row classes, and to use those throughout the code instead of direct property access. Otherwise, when we switch to Doctrine, all of the code will need to be refactored again anyway -- might as well get it done up front so we have less churn in the published code.

I'm sorry for the nuisance -- it's bad timing to add a lot of database tables while we're in the middle of a transition like this, but the goal is that this new interface-based approach will make any future changes much less disruptive.

demiankatz commented on module/VuFind/src/VuFind/AjaxHandler/NotificationsClose.php:
```diff
@@ -0,0 +1,104 @@
+        $broadcast_id = $params->fromPost('broadcast-id', $params->fromQuery('broadcast-id', null));
+
+        $session = $this->getSession();
+        if (!isset($session->closedBrodcasts)) {
```
Typo: I think `closedBrodcasts` should be `closedBroadcasts`.

demiankatz commented on module/VuFind/src/VuFind/AjaxHandler/NotificationsReorder.php:
```diff
@@ -0,0 +1,101 @@
+     *
+     * @var NotificationsPages
+     */
+    private $pagesTable;
```
As a general rule, our codebase avoids `private` properties and uses `protected` everywhere to more easily facilitate extending/overriding code. Might be worth replacing the private with protected everywhere, unless you feel there's a strong reason not to.

demiankatz commented on module/VuFind/src/VuFind/Form/NotificationsPagesForm.php:
```diff
@@ -0,0 +1,218 @@
+
+        foreach ($this->config['Notifications']['languages'] as $language) {
+            /*
+            $inputFilterSpecifications[] = [
```
Is all this commented-out code here for a reason?

demiankatz reviewed (changes requested):
Since this is still work in progress, I didn't do anything like a thorough review yet, but I spotted a few small things while quickly scrolling through the diffs.

It also looks like there's a minor conflict that can be resolved by merging the release-10.2 branch, deleting the conflicted `composer.lock`, running `composer install` and then adding/committing the regenerated `composer.lock`.

Additionally, there are some minor style problems that `composer fix` should automatically fix for you.

crhallberg commented:
Wanted to give my approval for this once Demian's review has been addressed!

demiankatz commented:
@jschultze, at today's Community Call, we decided to set the release date for 10.2 to Monday, June 23. In order to allow time for internationalization work, this means that any translations from this PR would need to be finalized before Monday, May 19. If you think it would be realistic to get this work done in time for the 6/23 release date, we can merge the translation strings early (so not ALL work needs to be done in just the next two weeks). However, if this timeline seems too rushed for you, we might want to defer back to the 11.0 release; with work on #2233 nearly complete, release 11.0 is probably not going to come too terribly long after 10.2, so that might not be an entirely bad thing.

Please let me know if you have thoughts/preferences, and if you expect to need more detailed support from me (in case I need to budget some of my time to assist).

jschultze commented:
@demiankatz Thanks for the timeline. Although I don't know how much more work will follow from a thorough review, we will surely try for the 10.2 release, as we really need to get this project out of our ToDo-List.

Regarding the translations: as far as I am aware, the administration of broadcasts and pages by now has understandable hints and finalized displayed texts. So if you don't think there's any more to add, we should be fine.

demiankatz committed:
Merge branch 'release-10.2' into dev_vufind_notifications_cms_rebase_10_2

demiankatz commented:
Thanks, @jschultze. I've just merged the release-10.2 branch into this branch and resolved conflicts so we can move forward with the work. It looks like right now, everything is broken -- I'm getting fatal `Class VuFind\\Db\\Row\\NotificationsPages contains 26 abstract methods and must therefore be declared abstract or implement the remaining methods` errors when I try to do anything. Do you need help getting the missing getters/setters added here? I'm a bit short on time this week, but I can try to spend a little time tomorrow getting things to a running state if necessary. It would be good to get this working again so we can review the language strings in context and make sure they're all finalized.

jschultze committed:
Added setters and getters.

jschultze committed:
Merge branch 'dev_vufind_notifications_cms_rebase_10_2' of https://github.com/jschultze/vufind_notifications_cms into dev_vufind_notifications_cms_rebase_10_2

jschultze commented:
@demiankatz Thanks for the merge. I added the getters and setters, and the code works again.

sturkel89 reviewed (changes requested):
I tested this PR and ran into a few problems.

### Language isssue
I created two notification pages in English (one that links to an external page and one that does not) and one in the German tab (regular content, not a link to an external page). When I view the list of three, I see my English titles, but there is a blank space for the German item.

![image](https://github.com/user-attachments/assets/1967534b-35f3-4fe4-a9b1-7cf3fd32f10a)

When I change the page language to German, I see the title for the German notification and for the English one that is NOT a link.

![image](https://github.com/user-attachments/assets/ed3c9727-c5e4-41d3-9e0a-a30b967cdbe3)

-----

### CSS display issue

In full screen, I only see about the first six characters of each notification page text. This is true if there is one visible or two. (Also, when the page is displayed in German, I don't see the item that is a link.)

![image](https://github.com/user-attachments/assets/e6023846-8738-4565-b12b-5d32039b9b77)

![image](https://github.com/user-attachments/assets/c4ce8ad4-31aa-4606-834f-2c95c05e5b80)

![image](https://github.com/user-attachments/assets/b494817a-e5a1-45d4-adc3-e5de8b52b969)

In medium screen widths, using the hamburger menu to view the menu items, you see more of the text:

![image](https://github.com/user-attachments/assets/c8c651a9-8577-4baa-b681-5c1f45902b25)

In mobile, you only see the first few characters:

![image](https://github.com/user-attachments/assets/2d47ffbf-36b4-46b8-a180-8a960eb7fea7)

---------------

### Error creating broadcast

When I try to create a notification broadcast, I get a "red box" error when I click **Save** in the form. 

When I turn on development mode and try again, the error that is reported is `Undefined variable $user`.

jschultze committed:
Fixed the display of pages on the pages list. Default language should be used if no local translation is available.

jschultze committed:
Fixed errors while creating new broadcasts.

jschultze committed:
Fixed the display of pages in the header. Default language should be used if no local translation is available.

jschultze committed:
Fixed the display of pages in the header.

jschultze commented:
@demiankatz @sturkel89 The following fixes are online:

- Empty entry in pages list
There was an error filling in the contents of the default language if no local language exist. This is fixed. But I do not have a conceptional idea how to deal with the other case: no default language, but other languages. In this case it might be easy to fill in German, but when more than two languages are configured and one page only exist in language 2 and one page only in language three, the display will get mixed quite a bit language-wise if the just keep searching for the first language that is available. Maybe a general warning "not available in default language" would do the job?

- Navigation
This problem was related, as the empty space was included for a link to a page that was missing the default language. I simply removed this link. The same problem with many mixed languages is the case, but a default text does not seem to be right in this place.

- Navigation elements width
I tried giving the element in the menu a bite more space.

Maybe this addition to the admin menu has to rely a bit on the understanding of it's limit through the user?

- $user error
This is fixed.

sturkel89 reviewed (changes requested):
I just reviewed the updated PR. There are some improvements, but still some issues.

### Language issues

On both the back end and the front end, content created in the German language tab is missing when the page language is set as English:

![image](https://github.com/user-attachments/assets/bdc848e4-866a-42d4-b8a8-6e9a90e9fab0)

vs. 

![image](https://github.com/user-attachments/assets/a746d8a7-13d2-4a40-9ba7-f9caf43e2f9b)

![image](https://github.com/user-attachments/assets/81c48b08-45b0-466f-bdad-3cd1e9ddd002)

The same exact thing is seen with broadcasts. Broadcasts created in either language are visible when page language is set as German, but when page language is set as English only the the English broadcasts appear. This is true on the back end and the front end.

![image](https://github.com/user-attachments/assets/248a14b2-85e7-4cbe-85cf-acc24415bc78)

However, the $user error is fixed.

Also, I still can't dismiss the broadcast using the X in the corner. It should be removed if it can't be functional.

Thanks for all your work on this very complex development project!

jschultze commented:
@demiankatz @sturkel89 I don't really know how to proceed with the fixes right now. The missing translations seem to be a conceptual problem, which I could read as either bug or feature. Regarding the closing functionality: we checked this multiple times with different browsers. Is the removal of the X for closing the final decision? If so, I would commit that change.

Will there still be a more thorough review of the code?

demiankatz commented:
@jschultze, I've been waiting for you and @sturkel89 to work through the functional issues before diving into the next technical review (and, unfortunately, my bandwidth for technical review has been limited by some other projects and priorities over here).

Regarding dismissing the broadcasts, I'm not sure why the "X" is not working -- but I think it should either work consistently or not be present. If you need help troubleshooting what is going wrong, I can certainly look into that further.

Regarding the translation issues, I think the key question is: will there ever be a circumstance in which a notification should only appear for one language and not another? If so, maybe the existing behavior is okay. But I think it's more likely that the desired behavior is that a notification should always display, and should use the most appropriate available language for its display. In other words, when editing a notification, you should have the option to provide translations in all supported languages. When the notification is displayed, it should try to display first the user's chosen language, then the site's default language, then perhaps whatever arbitrary language is provided. (Similar logic is used for displaying help screens -- I can find the relevant code if you need a reference).

Is that any help?

jschultze committed:
Changed `array` to `ArrayObject` in NotificationsBroadcasts methods and removed close button from broadcast templates.

jschultze committed:
Improved fallback logic for broadcasts and pages to handle multiple languages when translations are missing.

jschultze commented:
@demiankatz @sturkel89 

- The X to close the broadcasts is removed.
- The display of broadcasts and pages no searches all languages for the first one that has content in the required fields.

demiankatz commented on config/vufind/Notifications.yaml:
```diff
@@ -0,0 +1,24 @@
+Notifications:
+  # Activated languages.
+  # Add more as two character abbreviations.
+  languages:
```
Might it make sense for this setting to be optional, and default to whatever languages are activated in config.ini? That way, the user can narrow down the language list if they want to, but by default, they won't be forced to configure their language list in two different places.

demiankatz commented on config/vufind/permissionBehavior.ini:
```diff
@@ -91,3 +91,6 @@ deniedTemplateBehavior = showTemplate:error/loginForAccess
 ;deniedControllerBehavior = "exception::You are not logged in!"
+
+[notifications.Admin]
+deniedControllerBehavior = "showMessage:Access to notifications administration not granted"
```
Should this string be added to the language files for translation purposes?

demiankatz commented on languages/en.ini:
```diff
@@ -957,6 +965,27 @@ note_785_7 = "Merged with / Forms"
+notifications_pages = "Notifications Pages"
+notifications_pages_external_url = "External URL"
+notifications_pages_headline = "Headline"
+notifications_pages_is_external_url = "Link to exteral page"
```
Typo fix:

```suggestion
notifications_pages_is_external_url = "Link to external page"
```

demiankatz commented on module/VuFind/src/VuFind/View/Helper/Notifications/Notifications.php:
```diff
@@ -0,0 +1,293 @@
+ * @link     https://vufind.org Main Site
+ */
+
+namespace VuFind\View\Helper\Notifications;
```
We generally organize our view helpers on the basis of theme. Unless you're planning to have a group of notifications-related helpers, I'd suggest moving this to the VuFind\View\Helper\Root namespace.

demiankatz reviewed (changes requested):
Thanks for the progress, @jschultze. I've taken another look through all of this to see what the status is, and I've resolved a number of open conversation threads. However, it appears that there is still a review from April 25 that has not been addressed yet.

I've added a few new ideas and comments below.

I'm sorry that it hasn't been possible to get this all in shape in time for release 10.2. A couple of suggestions for a strategy to get this done:

1.) We should decide on an end goal -- do we want to continue targeting the 10.x code line, in order to make the code available in a form that existing users can install more quickly/easily, or should we shift gears to 11.x, so that we can focus on modernization and drop some legacy things (like the bootstrap3 theme) to reduce the number of files we're working on?

2.) Once we have a goal, would it be feasible to try to coordinate some focused time to work through the remaining code? I think it would be good to have a planned schedule for the review-coding-testing cycles to finish the work. Since there have been long pauses in development here, I think we're all losing time and energy refreshing our memories when we double back to the PR. It might be more efficient if we could either set up some kind of schedule (like "work on this for X hours once per week") or a concentrated sprint ("work on this daily for a week or two until it is done"). If we plan a predictable pattern for finishing the work, we could also potentially invite additional community participation to help.

Please let me know what you think, and what is feasible given your own time and resources.

Again, I'm sorry that this is taking so long -- but it's a large and complex feature that intersects with areas of the code that have been changing out of necessity, so it's just been a challenging thing to finish. I do believe we can get there, though! :-)

jschultze commented:
Thanks for the update, @demiankatz. We will have to plan internally when we can continue working on this pull request. As the big library congress in Germany is comming up in the following weeks, that might take some time. I'll get back to you as soon as I know more.

demiankatz commented:
I'm changing the base branch and milestone based on our discussion. Please let me know if you need help resolving conflicts!

demiankatz commented:
Reassigning milestone as per discussion on the 1 July Community Call.

ThoWagen commented:
This PR will be split into multiple smaller PRs to make reviewing easier. The first is #5048. So, this PR can be closed.
